lib/fine-types/src/Language/FineTypes/Typ/Gen.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Language.FineTypes.Typ.Gen where
+
+import Prelude
+
+import Data.Char (toUpper)
+import Data.List (nub)
+import Data.Traversable (forM)
+import Language.FineTypes.Typ
+    ( ConstructorName
+    , FieldName
+    , OpOne (..)
+    , OpTwo (..)
+    , Typ (Abstract, One, ProductN, SumN, Two, Var, Zero)
+    , TypConst (..)
+    , TypName
+    )
+import Test.QuickCheck
+    ( Gen
+    , elements
+    , listOf1
+    , oneof
+    , scale
+    , vectorOf
+    )
+
+-- | If the generated 'Typ' should be concrete or not. 'Concrete' will not contain
+-- 'Abstract' or 'Var' leaves
+data Concrete = Complete | Concrete
+
+patchNoData :: Concrete -> [Gen Typ] -> [Gen Typ]
+patchNoData Concrete = id
+patchNoData Complete =
+    (<>)
+        [ pure Abstract
+        , Var <$> genVarName
+        ]
+
+-- | Minimum depth of the generated 'Typ'. Shorter than depth branches are still
+-- possible if the actual  'Typ' is Zero or Abstract or Var
+newtype DepthGen = DepthGen Int
```

Comment:
I believe that using a plain type synonym is better value (type safety adds little value here, `instance Num` allows mixing with integer numerals, synonym avoids `GeneralizedNewtypeDeriving`), but I leave that up to you here.